File: src/main/java/com/refinedmods/refinedstorage/screen/widget/sidebutton/ExactModeSideButton.java
package com.refinedmods.refinedstorage.screen.widget.sidebutton;

import com.mojang.blaze3d.vertex.PoseStack;
import com.refinedmods.refinedstorage.api.util.IComparer;
import com.refinedmods.refinedstorage.screen.BaseScreen;
import com.refinedmods.refinedstorage.blockentity.data.BlockEntitySynchronizationManager;
import com.refinedmods.refinedstorage.blockentity.data.BlockEntitySynchronizationParameter;
import net.minecraft.ChatFormatting;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.resources.language.I18n;

public class ExactModeSideButton extends SideButton {
    private static final int MASK = IComparer.COMPARE_NBT;

    private final BlockEntitySynchronizationParameter<Integer, ?> parameter;

    public ExactModeSideButton(BaseScreen<?> screen, BlockEntitySynchronizationParameter<Integer, ?> parameter) {
        super(screen);

        this.parameter = parameter;
    }

    @Override
    protected String getSideButtonTooltip() {
        String tooltip = I18n.get("sidebutton.refinedstorage.exact_mode") + "\n" + ChatFormatting.GRAY;

        if ((parameter.getValue() & MASK) == MASK) {
            tooltip += I18n.get("sidebutton.refinedstorage.exact_mode.on");
        } else {
            tooltip += I18n.get("sidebutton.refinedstorage.exact_mode.off");
        }

        return tooltip;
    }

    @Override
    protected void renderButtonIcon(GuiGraphics graphics, int x, int y) {
        int ty = 16 * 12;
        int tx = (parameter.getValue() & MASK) == MASK ? 0 : 16;

        graphics.blit(BaseScreen.ICONS_TEXTURE, x, y, tx, ty, 16, 16);
    }

    @Override
    public void onPress() {
        BlockEntitySynchronizationManager.setParameter(parameter, parameter.getValue() ^ MASK);
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/screen/widget/sidebutton/ExactModeSideButton.kt
package com.refinedmods.refinedstorage.screen.widget.sidebutton

import com.mojang.blaze3d.matrix.MatrixStack
import com.refinedmods.refinedstorage.api.util.IComparer
import com.refinedmods.refinedstorage.screen.BaseScreen
import com.refinedmods.refinedstorage.tile.data.TileDataManager
import com.refinedmods.refinedstorage.tile.data.TileDataParameter
import net.minecraft.client.resources.I18n
import net.minecraft.util.text.TextFormatting

class ExactModeSideButton(screen: BaseScreen<*>, private val parameter: TileDataParameter<Int?, *>) : SideButton(screen) {
    override fun getTooltip(): String {
        var tooltip: String = I18n.format("sidebutton.refinedstorage.exact_mode").toString() + "\n" + TextFormatting.GRAY
        if (parameter.value!! and MASK == MASK) {
            tooltip += I18n.format("sidebutton.refinedstorage.exact_mode.on")
        } else {
            tooltip += I18n.format("sidebutton.refinedstorage.exact_mode.off")
        }
        return tooltip
    }

    override fun renderButtonIcon(matrixStack: MatrixStack?, x: Int, y: Int) {
        val ty = 16 * 12
        val tx = if (parameter.value!! and MASK == MASK) 0 else 16
        screen.blit(matrixStack, x, y, tx, ty, 16, 16)
    }

    fun onPress() {
        TileDataManager.setParameter(parameter, parameter.value!! xor MASK)
    }

    companion object {
        private const val MASK = IComparer.COMPARE_NBT
    }
}